Hydrate DNB search API responses with Data Hub company data if it exists

currycoder opened the pull request:
### Description of change

This PR aims to give DNB search results some context for the user by hydrating the results with Data Hub company data. Search results will be hydrated if the corresponding Company exists on Data Hub and if it has a duns number which matches the DNB duns number.

We have a requirement for the last interaction to show against each matched search result in the frontend, so this data is exposed using @reupen's pattern for getting the latest related entity to a model from a one to many relationship.  It's possible that we will need other company/related data to be exposed later, but for now this gives us a basis to get started with.

### Checklist

* [X] Has a new newsfragment been created? Check [changelog/README.rst](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.rst) for instructions
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [ ] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

mcbhenwood commented on datahub/dnb_api/queryset.py:
```diff
@@ -0,0 +1,31 @@
+
+def get_company_queryset():
+    """
+    Returns an annotated query set used by DNBCompnaySearchView.
```
"DNBCompanySearchView" :)

reupen commented on datahub/dnb_api/views.py:
```diff
@@ -26,15 +30,50 @@ class DNBCompanySearchView(APIView):
+            datahub_company = datahub_companies_by_duns.get(duns_number)
+            if datahub_company:
+                datahub_company_data = DNBMatchedCompanySerializer(datahub_company).data
+                hydrated_result['datahub_company'] = datahub_company_data
```
I'd try to define `hydrated_result` in one go and separate it from the conditional logic to make it a bit easier to follow. In other words, define `datahub_company_data` first and then `hydrated_result`.

currycoder replied:
Nice observation, thanks.

reupen commented:
There are some groups of users that don't have access to company interactions – will they have access to the D&B search?

Additionally, the search should require the `company.view_company` permission now it includes Data Hub company details. (There is a `HasPermissions` permission class that could be used for that.)

currycoder commented on datahub/dnb_api/views.py:
```diff
@@ -26,15 +30,50 @@ class DNBCompanySearchView(APIView):
             content_type=upstream_response.headers.get('content-type'),
         )
 
+    def _transpose_and_hydrate(self, dnb_results):
```
Good point.  It's more just changing the format of the results, I've changed it to `_format_and_hydrate()`

currycoder commented:
> There are some groups of users that don't have access to company interactions – will they have access to the D&B search?
> 
> Additionally, the D&B search view should probably require the `company.view_company` permission now it includes Data Hub company details. (There is a `HasPermissions` permission class that could be used for that.)

Thanks Reupen.  I've added in permissions for these parts.  As we discussed, I did look at `PermittedFieldsModelSerializer` for ensuring that `last_interaction` data was not shown to users that do not have permissions for interactions. Unfortunately it looked like it omitted the field entirely - for ease of frontend development (and to avoid us missing the edge case later) I've added my own code which just sets the data to `None` if the user does not have permission, while ensuring that the API still responds with a `last_interaction` key.

currycoder commented:
> I think there is an advantage of omitting the field: it more clearly differentiates between 'this company has no interactions and 'I can't tell you anything about the company's interactions'. But, the permission can also be used to work the latter out if needed so either should work.

I like the rationale here.  I think I'll go ahead and use the `PermittedFieldsModelSerializer` for the benefit of code reuse. Although using it does necessitate the serializer knowing about the request, which is at odds with your other observation.

reupen commented:
> Although using it does necessitate the serializer knowing about the request, which is at odds with your other observation.

I know :disappointed:

alixedi commented on datahub/dnb_api/views.py:
```diff
@@ -20,21 +25,92 @@ class DNBCompanySearchView(APIView):
+        }
+
+        """
+        duns_numbers = [result['duns_number'] for result in dnb_results]
```
I feel the code here can be a bit more succinct. I would propose something along the following lines to give you some ideas (prolly not all good ones):

```python
def get_matched_datahub_company(company):
    return DNBMatchedCompanySerializer(
        company, 
        context={'request': self.request})
    }.data

def format_dnb_companies(companies):
    return {
        company['duns_number']: {'dnb_company': company} 
        for company in companies
    }

def get_datahub_companies(dnb_numbers):
     return get_company_queryset().filter(duns_number__in=dnb_numbers)

def format_datahub_companies(companies):
    return {
        company.duns_number: {'datahub_company': get_matched_datahub_company(company)
        for company in companies
    }

def format_and_hydrate(self, dnb_results):
    dnb_companies = format_dnb_companies(dnb_results)
    datahub_companies = format_datahub_companies(
        get_datahub_companies(dnb_companies.keys())
    )

    return {
        # All companies found in DNB 
        **dnb_companies, 
        # Override ones that were matched to DataHub with richer matched objects
        **{
            k: {**datahub_companies[k], **dnb_companies[k]} 
            for k in data hub_companies.keys()
        }
    }.values()
```

Essentially, trying to break it down into small testable units.

currycoder replied:
Honestly, I feel like the specific suggestion here is a little over the top - to me, it replaces complexity in the original solution with new complexity from dealing with intermediate dictionaries and then merging them in a complex comprehension.

I do think there's a good point about breaking the method down in to separate chunks for readability/maintainability though.  I've broken parts out of the original method in to helpers with the hope that the overall `_format_and_hydrate` method is more readable.

reupen commented on datahub/dnb_api/views.py:
```diff
@@ -20,21 +25,98 @@ class DNBCompanySearchView(APIView):
+                datahub_company,
+                context={'request': self.request},
+            ).data
+        else:
```
This `else:` can be removed (and `return None` brought up an indentation level) now.

reupen commented on datahub/dnb_api/test/test_views.py:
```diff
@@ -18,60 +26,141 @@ def dnb_company_search_feature_flag():
 
         assert response.status_code == response_status_code
-        assert response.content == response_content
+        assert json.loads(response.content) == response_data
```
Are we able to use `response.json()` (and in similar lines elsewhere)?

reupen commented on datahub/dnb_api/views.py:
```diff
@@ -20,21 +25,97 @@ class DNBCompanySearchView(APIView):
+
+        hydrated_results = []
+
+        for dnb_result in dnb_results:
```
It might be worth having a go at moving the loop body to a separate function and using a list comprehension – I'll leave it with you to have a look.

elcct replied:
I think this is a good suggestion to move the body to a separate function.

reupen commented on datahub/dnb_api/test/test_views.py:
```diff
@@ -18,60 +26,141 @@ def dnb_company_search_feature_flag():
+    # Company with two interactions
+    company = CompanyFactory(duns_number='7654321', id='6083b732-b07a-42d6-ada4-c99999999999')
+
+    interaction_date = make_aware(
```
Elsewhere we pass `tzinfo` explicitly instead of using `make_aware()` – it would be good to keep it consistent.

reupen commented on datahub/dnb_api/test/test_views.py:
```diff
@@ -18,60 +26,141 @@ def dnb_company_search_feature_flag():
+        subject='Meeting with Joe Bloggs',
+        company=company,
+    )
+    latest_interaction.created_on = interaction_date
```
Although this should work, I think I would use `freeze_time()` since overwriting `created_on` wouldn't normally be done.

reupen commented on datahub/dnb_api/test/test_views.py:
```diff
@@ -18,60 +26,141 @@ def dnb_company_search_feature_flag():
+        subject='Meeting with John Smith',
+        company=company,
+    )
+    older_interaction.created_on = older_interaction_date
```
This change doesn't seem to be saved (but see previous comment too).
